### DISCUSSION TITLE
Ability to create own ImageProcessor

### DISCUSSION THREAD

**Discussion by UkooLabs** [SixLabors/ImageSharp#1366] [Q&A]
Hi,

I started to try to create my own ImageProcessor however i got stuck on the OnFrameApply code.

It appears there is no way of cloning the source Image nor applying changes to it. (due to internals)

I also tried a test with no luck, the following...

```
        protected override void OnFrameApply(ImageFrame<TPixel> source)
        { 
            for (var x = SourceRectangle.Left; x < SourceRectangle.Right; x++)
            {
                for (var y = SourceRectangle.Top; y < SourceRectangle.Bottom; y++)
                {
                    var temp = source[x, y].ToVector4();
                    temp.X = 1;
                    source[x, y].FromVector4(temp);
                }
            }
        }
```

I'm currently in need to write a fast blur routine soon based on http://blog.ivank.net/fastest-gaussian-blur.html and looks like I maybe blocked on doing something truly fast.

Please confirm if I am doing something wrong or not

If does not exist publicly the ability to copy and apply pixel regions to Image/ImageFrames would be great

**Comment by UkooLabs** [comment]
Ok my test was wrong should of been something like (which works)...

```
        protected override void OnFrameApply(ImageFrame<TPixel> source)
        {
            for (var x = SourceRectangle.Left; x < SourceRectangle.Right; x++)
            {
                for (var y = SourceRectangle.Top; y < SourceRectangle.Bottom; y++)
                {
                    var pixel = new TPixel();
                    var vector4 = source[x, y].ToVector4();
                    vector4.X = 1;
                    pixel.FromVector4(vector4);
                    source[x, y] = pixel;
                }
            }
        }
```

how ever the ability to copy and replace pixel regions still stands as a great to have

**Reply by JimBobSquarePants** [reply]
You're missing access to `Buffer2D<TPixel>` aren't you? Not sure why we don't expose it. 

Will have to check the code to see if we can, I'll do that tomorrow. 

On the note of a fast `Gaussian`. Ours is not slow since we use a separable algorithm but I've actually been wanting to update our `BloxBlur` implementation to use the travelling average approach for a while just never gotten round to it since the current implementation works. It would be nice if you could upgrade our implementation and everyone could benefit.

**Reply by UkooLabs** [reply]
@JimBobSquarePants 

I have access to..

using Buffer2D<TPixel> targetPixels = this.Configuration.MemoryAllocator.Allocate2D<TPixel>(source.Width, source.Height);

however the following I dont...

source.**CopyTo**(targetPixels);
source.**PixelBuffer**
Buffer2D<TPixel>.**SwapOrCopyContent**(source.**PixelBuffer**, targetPixels);

As a note the gausian blur using a 1600x1027 image using 100px radius takes 15,000ms where the one i linked and have ported to ImageSharp takes only 340ms and looks identical to gausian compared to boxblur which doesnt look so nice. Here is the ImageProcessor...

https://gist.github.com/UkooLabs/bb46e7ad4219bab46968409b74d6018f

**Reply by JimBobSquarePants** [reply]
Yeah, it was `PixelBuffer` I was thinking of. 

Fast Guassian is essentially BoxBlur ran 3 times.  Our BoxBlur could be optimized to use the travelling average algorithm rather than the separable approach we use now.
